### PR TITLE
filebeat: ID for filestream is required

### LINKED ISSUE #30061: Filestream input duplicating events after every restart
Under some circumstances the filestream input is processing all the events after every restart.

For example the following configuration works fine in a Filebeat running on Kubernetes (static input, no autodiscover):

```
    - type: filestream
      enabled: true
      paths:
       - /var/log/k8sapps/myapp/*.log
      fields:
       app.name: "myapp"
      fields_under_root: true
```

But if we add a second input (actually from same disk) then Filebeat sends everything after every restart:

```
    - type: filestream
      enabled: true
      paths:
       - /var/log/k8sapps/myapp/*.log
      fields:
       app.name: "myapp"
      fields_under_root: true

    - type: filestream
      enabled: true
      paths:
       - /var/log/k8sapps/secondapp/*.log
      fields:
       app.name: "secondapp"
      fields_under_root: true
```

I've tried to apply `file_identity.inode_marker.path: /var/log/.filebeat-marker` but the result is the same, and with a single input all works as expected.
The inodes of the files do not change after every restart and the volume UUID i don't know because it's not reported by `lsblk` (checked from the filebeat container itself).

Doc reference: https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-filestream.html#filestream-file-identity

### PR DESCRIPTION
**Pull request by belimawr**
## What does this PR do?
It makes mandatory to configure an ID to all filestream inputs. If an ID is added to a existing, working, configuration, no data is duplicated.

### How does it work
When a filestream input does not have an ID, we set the default `.global` ID at runtime and this ID is used to identify which files in the registry belong to each input.

Now that we're making the IDs mandatory, we need a way to migrate those entries in the registry to the newly set ID. Ideally this will happen only once when the users update to this new version of Filebeat.

They way we do it is by looking the registry for files that have their path matching the input but have the input ID as `.global`, those entries are then copied with the new key (which uses the input ID) so Filebeat can continue its operation without duplicating data.

Most of the logic is the same of the already existing method `UpdateIdentifiers` from the `sourceStore` but for this PR it was also needed to update the in-memory store right away. Also borrowing from `UpdateIdentifiers` usage, this is done during the file prospector initialisation.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

~~## Author's Checklist~~
## How to test this PR locally
1. Create a simple config that reads a static file using `filestream`, do not set an ID for this input
2. Run Filebeat (a non patched version) and wait all events to be ingested
3. Stop Filebeat
4. Add an ID to your filestream input
5. Start a patched version of Filebeat
6. No new events should be ingested
7. The registry should have its entries updated with the new input ID 

## Related issues

- Closes #30061

~~## Use cases~~
~~## Screenshots~~
~~## Logs~~

### CODE REVIEW

File: filebeat/input/filestream/internal/input-logfile/store.go
Comment (cmacknz) [inline filebeat/input/filestream/internal/input-logfile/store.go:231]: Why would this fail?

File: filebeat/input/filestream/internal/input-logfile/store.go
Comment (cmacknz) [inline filebeat/input/filestream/internal/input-logfile/store.go:238]: I don't understand this code very well, but calling `Unlock` before a `continue` here and then again when the function returns below seems suspicious to me.

**Comment by cmacknz** [comment]
Could you add a small explanation of what your proposed change is doing? I suspect this is a tricky part of the code and some more explanation/comments would help make it easier to review.

File: filebeat/docs/inputs/input-filestream-file-options.asciidoc
Comment (kvch) [inline filebeat/docs/inputs/input-filestream-file-options.asciidoc:24]: Please add a warning about changing existing IDs cause data duplication.

File: filebeat/input/filestream/internal/input-logfile/store.go
Comment (kvch) [reply filebeat/input/filestream/internal/input-logfile/store.go:231]: If something else has already locked the resource `TryLock` returns `false` immediately. When `FixUpdateIdentifiers` is called in `Init`, nothing __should__ be locking the resources in theory. But better to be safe than sorry.

@belimawr I am rather curious why you decided to use `TryLock` instead of `Lock`? If the resource is locked by something else, shouldn't we wait until we get access to fix its ID?

File: filebeat/input/filestream/internal/input-logfile/manager.go
Comment (kvch) [inline filebeat/input/filestream/internal/input-logfile/manager.go:198]: Why are you cleaning during prospector initialization? I think we can do it earlier when the manager is initialized: https://github.com/elastic/beats/blob/main/filebeat/input/filestream/internal/input-logfile/manager.go#L94 or https://github.com/elastic/beats/blob/main/filebeat/input/filestream/internal/input-logfile/manager.go#L118

File: filebeat/input/filestream/internal/input-logfile/store.go
Comment (cmacknz) [reply filebeat/input/filestream/internal/input-logfile/store.go:231]: Thanks, I think what I really wanted to ask is why we use `TryLock` instead of `Lock`. Why is it OK to fail to obtain the lock and avoid executing the code below?

Comment (belimawr) [reply filebeat/input/filestream/internal/input-logfile/store.go:231]: You're totally right @kvch ! I had copied `UpdateIdentifiers` and then did some small modifications to fit my use case without putting much thought on the existing code. I'll fix that and double check the rest of the code.

It's not ok to fail to obtain the lock, we should wait.

File: filebeat/input/filestream/internal/input-logfile/store.go
Comment (belimawr) [reply filebeat/input/filestream/internal/input-logfile/store.go:238]: The idea here is if the `newKey` is already present on `s.store.ephemeralStore.table` we won't replace it, so we unlock the resource and move on to the next iteration of the `for` loop.

If `newKey` is not present on `s.store.ephemeralStore.table`, then we add it and then unlock at the very end of the `for` loop iteration.

File: filebeat/input/filestream/internal/input-logfile/manager.go
Comment (belimawr) [reply filebeat/input/filestream/internal/input-logfile/manager.go:198]: Because that is where the registry is cleaned and also the "update identifiers" runs. It looked like a similar operation to me. Also, most of the data needed to implement this logic is already available there.

I'll look into those other places you mentioned as well.

Comment (belimawr) [reply filebeat/input/filestream/internal/input-logfile/manager.go:198]: I took a look at those methods, but they do not have access to the input IDs, now the "configuration yaml". The first place I saw with access to the input IDs is the `Create` method: https://github.com/elastic/beats/blob/9e6536d964f0135940b9c416bcb1c0a41af99478/filebeat/input/filestream/internal/input-logfile/manager.go#L157-L169


But I also need the files information from the prospector, so the `prospector.Init` still looks like the best place for me.

File: filebeat/docs/inputs/input-filestream-file-options.asciidoc
Comment (belimawr) [reply filebeat/docs/inputs/input-filestream-file-options.asciidoc:24]: Done on c1492d8ec9

File: filebeat/input/filestream/internal/input-logfile/store.go
Comment (belimawr) [reply filebeat/input/filestream/internal/input-logfile/store.go]: Fixed on c1492d8ec9

File: filebeat/input/filestream/internal/input-logfile/store.go
Comment (cmacknz) [reply filebeat/input/filestream/internal/input-logfile/store.go:232]: I just misread the code here, thanks for clarifying.

**Comment by jlind23** [comment]
@belimawr could you please backport it to 7.17.X and all versions above?

**Comment by belimawr** [comment]
@cmacknz @kvch it's ready for review again. Let me know what you think of the python integration tests. I'm wondering if I should add more test cases or not.

File: filebeat/tests/system/test_input.py
Comment (cmacknz) [inline filebeat/tests/system/test_input.py:33]:
```suggestion
        with open(testfile, 'w') as file:
           offset = 0  # keep count of the bytes written
           for n in range(0, 5):
               offset += file.write("hello world\n")
```

Using `with` will automatically close the file when exiting the scope of the `with`.

File: filebeat/tests/system/test_input.py
Comment (cmacknz) [inline filebeat/tests/system/test_input.py:73]:
```suggestion

```
I don't think this try/except block does anything

File: filebeat/tests/system/test_input.py
Comment (cmacknz) [inline filebeat/tests/system/test_input.py:65]: What exception would be getting here? `RangeError` or similar if there is no `fixed_entry`? Can't you just check or assert on `len(entries)` first?

File: filebeat/input/filestream/internal/input-logfile/manager.go
Comment (cmacknz) [inline filebeat/input/filestream/internal/input-logfile/manager.go:167]: Will this prevent filebeat from starting if it was configured to use filestream inputs without an ID? This would be a breaking change wouldn't it? I'm not sure we want to do this.

Consider that there are customers with 1000+ agents that use filebeat, and they'll all just fail to start with this correct? If you are running with fleet or a large k8s deployment seeing this error is not as simple as reading one log file or some output in the terminal.

File: filebeat/input/filestream/internal/input-logfile/store.go
Comment (cmacknz) [inline filebeat/input/filestream/internal/input-logfile/store.go:216]:
```suggestion
// because it is meant to migrate inputs without IDs to inputs with IDs without
```

File: filebeat/input/filestream/internal/input-logfile/store.go
Comment (cmacknz) [inline filebeat/input/filestream/internal/input-logfile/store.go:240]: > 			// This might lead to duplicates if configurations are adapted
> 			// for inputs with the same ID are changed.

I'm not sure what this comment is telling me. Can you clarify it?

File: filebeat/input/filestream/internal/input-logfile/store.go
Comment (cmacknz) [inline filebeat/input/filestream/internal/input-logfile/store.go:251]:
```suggestion
			s.store.UpdateTTL(res, 0) // aka delete. See store.remove for details
```

File: filebeat/tests/system/test_input.py
Comment (belimawr) [reply filebeat/tests/system/test_input.py:70]: That's some leftovers from debugging the tests that, were breaking on Windows. Checking the length is definitely better.

I'll do some clean up once I get this test passing on Windows.

File: filebeat/input/filestream/internal/input-logfile/manager.go
Comment (belimawr) [reply filebeat/input/filestream/internal/input-logfile/manager.go:167]: Yes, Filebeat will fail to start and this is a breaking change, but not requiring an ID leads to data duplication with multiple inputs. I understood we were fine with this breaking change. Maybe we should discuss it a bit more.

I see 3 scenarios:
* Users run Filebeat with multiple filestream inputs without ID and have not noticed the data duplication on restart, it's a silent bug. This change will make it explicit. We can improve the error message.
* Users run standalone Filebeat and have set IDs to avoid the data duplication.
* Filebeat runs under Elastic-Agent, we're going to release together Fleet UI and Elastic-Agent to ensure all filestream inputs have an ID.

File: filebeat/input/filestream/internal/input-logfile/store.go
Comment (belimawr) [reply filebeat/input/filestream/internal/input-logfile/store.go:240]: Neither do I :/.
This was a copy and paste from `UpdateIdentifiers`: https://github.com/elastic/beats/blob/4a71e5e7eb8faa711785aee434a2aadd248439e0/filebeat/input/filestream/internal/input-logfile/store.go#L236-L244

I could not use `UpdateIdentifiers` for this feature because it wasn't updating the in-memory registry (`s.store.ephemeralStore`), so I copied it and did the small modifications I needed.

File: filebeat/input/filestream/internal/input-logfile/store.go
Comment (belimawr) [reply filebeat/input/filestream/internal/input-logfile/store.go:251]: Done

File: filebeat/tests/system/test_input.py
Comment (belimawr) [reply filebeat/tests/system/test_input.py:78]: Yep, that was part of my debugging-on-CI changes. I fixed it now.

File: filebeat/input/filestream/internal/input-logfile/store.go
Comment (belimawr) [reply filebeat/input/filestream/internal/input-logfile/store.go:240]: @kvch could you help us out here?

**Comment by belimawr** [comment]
@cmacknz I implemented the changes you suggested on [898e092](https://github.com/elastic/beats/pull/30717/commits/898e092332360c75ee7532cdf2bd8a598d119d09) and resolved the comments.

I finally got the tests passing on my Windows environment, once CI is happy I'll do a last code clean up and request a re-review.

File: filebeat/input/filestream/internal/input-logfile/manager.go
Comment (cmacknz) [reply filebeat/input/filestream/internal/input-logfile/manager.go:167]: I remember discussing that we should warn before completing deprecating filestream inputs without IDs but not much more than that. Hopefully I did not have a different opinion previously :)

The current situation results in duplication, which is an inconvenience. Preventing filebeat from starting could result in data loss (because of log rotation, for example). This is a worse outcome to me. We have escalated a bug into an outage.

I don't think we should make a breaking change here. Especially we should not make a breaking change in a minor release after a major full of breaking changes. 

I think the changes in this PR have value without breaking existing configurations. We should strongly recommend and obviously communicate the need to add IDs without breaking anything. 

I would suggest we consider automatically migrating configurations to have IDs if we can, but I don't want to derail the conversation too much with alternative solutions. We can ship this change as is without the breaking part of it.

Comment (belimawr) [reply filebeat/input/filestream/internal/input-logfile/manager.go:167]: > I remember discussing that we should warn before completing deprecating filestream inputs without IDs but not much more than that. Hopefully I did not have a different opinion previously :)

That's what I remember as well, there is a [PR merged](https://github.com/elastic/beats/pull/30401) with the warnings (I'm not completely sure it has been released already).

> The current situation results in duplication, which is an inconvenience. Preventing filebeat from starting could result in data loss (because of log rotation, for example). This is a worse outcome to me. We have escalated a bug into an outage.

That's a very good point.

> I don't think we should make a breaking change here. Especially we should not make a breaking change in a minor release after a major full of breaking changes.

I see, it makes sense. It's tricky because that's the only way to prevent the bug :/

> I think the changes in this PR have value without breaking existing configurations. We should strongly recommend and obviously communicate the need to add IDs without breaking anything. We would need to do that anyway.

So that would apply to the IDs duplication as well, right? I'll change everything to log error then.

> I would suggest we consider automatically migrating configurations to have IDs if we can as a follow up. I don't want to derail the conversation too much with alternative solutions. We can ship this change as is without the breaking part of it.

I believe that means re-writing the configuration file, otherwise the IDs will not persist neither be stable. Anyway I agree we can have this discussion later.

Comment (cmacknz) [reply filebeat/input/filestream/internal/input-logfile/manager.go:167]: > So that would apply to the IDs duplication as well, right? I'll change everything to log error then.

Ugh, yes I suppose if there are duplicated IDs out there now then preventing startup for them is a breaking change too.

Let's upgrade the warnings to errors as you suggested for these cases, and start a follow up conversation about ways to automatically fix this problem.

At a minimum, when we release this we should be able to have support resolve filestream duplication related cases without our input after this change. We will have to make sure we put this fix in their knowledge-base, and make it really obvious in the release notes.

**Review by cmacknz — CHANGES_REQUESTED** [review]
Requesting a change to remove the breaking behaviour as discussed.

File: CHANGELOG.next.asciidoc
Comment (cmacknz) [inline CHANGELOG.next.asciidoc:64]: We will have to reword this (and remove the breaking change note above). I would be as detailed as possible about what the issue is and how to fix it.

File: filebeat/input/filestream/internal/input-logfile/manager.go
Comment (kvch) [reply filebeat/input/filestream/internal/input-logfile/manager.go:211]: > Because that is where the registry is cleaned and also the "update identifiers" runs.

Updating identifiers is the responsibility of the prospector. It should only update state identifiers if it owns the given states. However, cleaning up legacy states does not belong to any specific prospector. It is the responsibility of the filestream manager to clean and maintain all filestream states.

> they do not have access to the input IDs

Why do we need access to input IDs? We know that we have to clean up the states that belong to `.global` ID.

File: filebeat/input/filestream/internal/input-logfile/store.go
Comment (kvch) [reply filebeat/input/filestream/internal/input-logfile/store.go:240]: The last sentence does not make much sense. :facepalm: :rofl: 

What I wanted to explain is just a corner case, it does not impact us because we do not update identifiers during data collection at the moment.

However, if identifiers are updated during data collection, it is possible that there are state updates that haven't been ACKed (by the output) when the `copyWithNewKey` is called. `copyWithNewKey` copies the existing state information with a new ID. So if there are outstanding updates to the state with the old ID, those updates will not be persisted to the newly created state with the new ID. So when Filebeat picks up the state with the new ID for the same source, the data will be duplicated because the updates of the old state are "lost".

Example for data duplication: Filebeat is collecting a `test.log` file with the ID of `filestream::test.log` and the last ACKed offset is 23. If the file is being harvested and some events are not yet been ACKed by Elasticsearch until the offset of 50. If Filebeat updates the ID of the state from `filestream::test.log` to `filestream::1234` the offset will be 23 in the registry. The data from bytes 23 to 50 in the file `test.log` is going to be resent to Elasticsearch regardless of the previous ACK because the ACK did not make it to Filebeat in time.

But as I mentioned above at the moment we do not update identifiers during data collection.

**Comment by cmacknz** [comment]
> They way we do it is by looking the registry for files that have their path matching the input but have the input ID as .global, those entries are then copied with the new key (which uses the input ID) so Filebeat can continue its operation without duplicating data.

Can there be path collisions between filestream inputs? If so does this new logic still work?

I am trying to think of edge cases I didn't see in the tests :)

File: filebeat/input/filestream/internal/input-logfile/manager.go
Comment (kvch) [reply filebeat/input/filestream/internal/input-logfile/manager.go:211]: Nevermind, I've reread you code. You are correct.

**Comment by belimawr** [comment]
> Can there be path collisions between filestream inputs? If so does this new logic still work?

I don't think so. As far as I know, we don't support multiple inputs reading from the same file. However I did not find it explicitly stated on our documentation for `filestream`. For the `log` input, that's very clear.

> I am trying to think of edge cases I didn't see in the tests :)

I didn't manage to invest much time thinking about them. But there are some things I still would like to test (at least locally):
-  Different file identity options
- Cloud provides/network share. For those I don't know how easy it is to reproduce the normal issues they can bring. However because I'm just manipulating the registry at start up, I don't believe it will play a big role.

I did test multiple inputs, it works fine.

File: CHANGELOG.next.asciidoc
Comment (belimawr) [reply CHANGELOG.next.asciidoc]: What do you think of the new version? Technically it's not a full bug fix because the users can still make it happen by not setting IDs despite all the warnings in the logs.

File: CHANGELOG.next.asciidoc
Comment (cmacknz) [inline CHANGELOG.next.asciidoc:62]: Close, I don't think we have mitigated it, but we have allowed mitigating it with a config change. What about something like:

```suggestion
- Allow fixing data duplication on restart when filestream inputs did not have an ID set. Setting an ID for filestream
  inputs will fix the problem without generating any data duplication{issue}30061[30061]
```

File: filebeat/input/filestream/internal/input-logfile/manager.go
Comment (cmacknz) [inline filebeat/input/filestream/internal/input-logfile/manager.go:177]:
```suggestion
			" duplication, please add an ID and restart Filebeat")
```

**Review by cmacknz — APPROVED** [review]
I don't have anymore concerns. Wait for approval from @kvch before merging.

File: filebeat/input/filestream/environment_test.go
Comment (kvch) [inline filebeat/input/filestream/environment_test.go:90]: Nit: this function could be called from `mustCreateInput`:

```golang
func (e *inputTestingEnvironment) mustCreateInput(config map[string]interface{}) v2.Input {
	e.t.Helper()
	inp, err := createInput(config)
	if err != nil {
		e.t.Fatalf("failed to create input using manager: %+v", err)
	}
	return inp
}
```

Comment (kvch) [reply filebeat/input/filestream/environment_test.go:90]: Also, I think this function could be marked as helper as well.

File: filebeat/input/filestream/environment_test.go
Comment (kvch) [inline filebeat/input/filestream/environment_test.go:215]: The header comment seems incorrect here. But it's not new, so you can ignore this problem.

**Review by kvch — APPROVED** [review]
I approve, but for the record I don't really like the new interface of `prospector.Init`. I am worried that every time we add a new maintainer/cleaner step to the registry, we will add more and more parameters. I would prefer if we could add callbacks or maybe add transformation steps to the `ProspectorCleaner`, so cleaner details and dependencies could remain hidden.

**Comment by kvch** [comment]
In the meantime I've come up with an idea. We could create a new special `sourceStore` that also has a sourceIdentifier for transforming the old IDs. So we would not have to pass a sourceIdentifier in the `Init` function. WDYT?

```golang
type transformerStore struct {
    sourceStore
    idTransformer sourceIdentifier
}
```

File: filebeat/input/filestream/environment_test.go
Comment (belimawr) [reply filebeat/input/filestream/environment_test.go:90]: In this case I needed one that would return the error from the input creation, hence I didn't use the already existing `mustCreateInput`.

I didn't mark it  as a helper because it does not call any method on `testing.T`. But there would be no harm in doing so.

File: filebeat/input/filestream/environment_test.go
Comment (belimawr) [reply filebeat/input/filestream/environment_test.go:215]: indeed. Well CI is red anyways, so I might as well fix this ;)

**Comment by belimawr** [comment]
> In the meantime I've come up with an idea. We could create a new special `sourceStore` that also has a sourceIdentifier for transforming the old IDs. So we would not have to pass a sourceIdentifier in the `Init` function. WDYT?
> 
> ```go
> type transformerStore struct {
>     sourceStore
>     idTransformer sourceIdentifier
> }
> ```

Then `Prospector.Init` would only receive a list of `ProspectorCleaner`, and `transformerStore` would also implement ProspectorCleaner, is that it?

I think it makes sense and keeps `Prospector.Init` clean.

**Comment by belimawr** [comment]
I fixed all lint issues, rebased onto `main` and force pushed.

File: .golangci.yml
Comment (belimawr) [inline .golangci.yml:154]: At least one line was failing in 3+ linters, but they're reported one at a time... It looks too strict to a repo like Beats.

File: .golangci.yml
Comment (belimawr) [inline .golangci.yml:123]: Now it matches the `.go-version` file/version.

**Comment by belimawr** [comment]
@rdner I did some small modifications in the lint config to reduce noisy, what do you think?

File: dev-tools/templates/.golangci.yml
Comment (rdner) [inline dev-tools/templates/.golangci.yml:20]:
```suggestion
      text: "ST1003:" # Poorly chosen identifier
```

File: .golangci.yml
Comment (rdner) [inline .golangci.yml:23]:
```suggestion
      text: "ST1003:" # Poorly chosen identifier
```

**Review by rdner — COMMENTED** [review]
I think we should leave an explanation comment for what we exclude, so people would not have to look it up all the time.

source https://staticcheck.io/docs/checks/

File: .golangci.yml
Comment (belimawr) [reply .golangci.yml:23]: I had it at the top, but I agree the yaml structure didn't help:
```
    # Exclude package name contains '-' issue because we have at least one package with
    # it on its name.
```

What about having it like that:
```yaml
    # Exclude package name contains '-' issue because we have at least one poorly chosen
    # package with'-' it on its name.
    - text: "ST1003:" 
      linters:
        - stylecheck
```

File: dev-tools/templates/.golangci.yml
Comment (belimawr) [reply dev-tools/templates/.golangci.yml:20]: I'll mark this one as resolved because fixing the template will also fix this file.

File: .golangci.yml
Comment (rdner) [reply .golangci.yml:23]: I missed that, I thought that comment was not related to the particular exception but now I see it's a list and the comment is on top of the item.

**Review by rdner — APPROVED** [review]

File: .golangci.yml
Comment (belimawr) [reply .golangci.yml:23]: The original yaml snipped I copied from is pretty confusing, but after taking some time to decipher it, I came up with something I hope is more readable.

**Comment by belimawr** [comment]
CI is failing: `[2022-03-24T15:26:18.127Z] Error: exec: "pytest": executable file not found in %PATH%` :sob:

**Pull request merged**
